File: tests/fixtures/java/Buggy.java
// EXPECTED FINDINGS:
// - line 19: java-hardcoded-password — security — Hardcoded password literal
// - line 22: java-sqli-statement-concat — security — Statement.executeQuery with concat
// - line 26: java-runtime-exec-user-input — security — Runtime.exec with variable
// - line 30: java-xxe-documentbuilder — security — DocumentBuilderFactory unsafe
// - line 34: java-weak-crypto-md5 — security — MessageDigest MD5
// - line 39: java-deserialization-readobject — security — ObjectInputStream.readObject
// - line 43: java-resource-not-closed — bug — FileInputStream not in try-with-resources
// - line 51: java-empty-catch — logic — Empty catch block
// - line 56: java-system-out-println-prod — quality — System.out.println

import java.io.*;
import java.sql.*;
import javax.xml.parsers.*;
import java.security.MessageDigest;

public class Buggy {

    public static final String password = "topsecretpassword";

    public static ResultSet lookupUser(Statement stmt, String name) throws SQLException {
        return stmt.executeQuery("SELECT * FROM users WHERE name = '" + name + "'");
    }

    public static Process runCmd(String cmd) throws IOException {
        return Runtime.getRuntime().exec(cmd);
    }

    public static DocumentBuilderFactory newDocFactory() {
        return DocumentBuilderFactory.newInstance();
    }

    public static byte[] fingerprint(byte[] data) throws Exception {
        return MessageDigest.getInstance("MD5").digest(data);
    }

    public static Object loadSession(byte[] blob) throws Exception {
        ObjectInputStream ois = new ObjectInputStream(new ByteArrayInputStream(blob));
        return ois.readObject();
    }

    public static String readFile(String path) throws IOException {
        FileInputStream fis = new FileInputStream(path);
        byte[] buf = fis.readAllBytes();
        return new String(buf);
    }

    public static void safelyParse(String s) {
        try {
            Integer.parseInt(s);
        } catch (NumberFormatException e) {
        }
    }

    public static void debugTrace(String message) {
        System.out.println("trace: " + message);
    }
}

File: tests/fixtures/kotlin/Buggy.kt
// EXPECTED FINDINGS:
// - line 17: kotlin-hardcoded-secret — security — Hardcoded SECRET constant
// - line 20: kotlin-weak-crypto-md5 — security — MessageDigest MD5
// - line 24: kotlin-runtime-exec — security — Runtime.exec call
// - line 28: kotlin-http-cleartext — security — Cleartext HTTP URL literal
// - line 32: kotlin-webview-js-enabled — security — WebView javaScriptEnabled = true
// - line 36: kotlin-double-bang — bug — Double-bang null assertion
// - line 42: kotlin-empty-catch — bug — Empty catch block
// - line 46: kotlin-todo-fn-call — bug — TODO() call left in code
// - line 50: kotlin-println-prod — quality — println in production

import android.webkit.WebView
import java.security.MessageDigest

object Buggy {

    const val SECRET = "supersecretlivekey"

    fun fingerprint(data: ByteArray): ByteArray {
        return MessageDigest.getInstance("MD5").digest(data)
    }

    fun runCmd(cmd: String): Process {
        return Runtime.getRuntime().exec(cmd)
    }

    fun apiBase(): String {
        return "http://"
    }

    fun configureWebView(wv: WebView) {
        wv.settings.javaScriptEnabled = true
    }

    fun firstChar(value: String?): Char {
        return value!!.first()
    }

    fun safelyParse(s: String): Int {
        return try {
            s.toInt()
        } catch (e: Exception) {}
    }

    fun unfinished(): String {
        TODO("not yet implemented")
    }

    fun debugTrace(value: Any) {
        println("trace: $value")
    }
}
